### PR TITLE
feature: cri manager skeleton

### LINKED ISSUE #420: [enhancement request] CRI manager to integrate pouch with kubernetes
## What is the scope of CRI manager
* GRPC server : connect to kubelet and receive CRI requests

* API transform : use pouch's container & image manager to handle all kinds of CRI requests

* Pod support : pod's lifecycle and metadata store

* CNI integration: the network of containers which belong to kubernetes should be managed by CNI other than the pouch's network manager

## What is not in the scope of CRI manager
* CRI manager should never interact directly with containerd. It is the job of pouch's container & image manager to implement all the details of containers & images processing

* All metadata, except for the sandbox should not be managed by CRI manager

The above is the coarse-grained division of CRI manager's scope, we may have to refine it in the future. And the next part of this issue is used to track the progress of CRI manager.

### Basic framework
- [x] [P0]Skeleton of CRI manager #426 

### Lifecycle
- [x] Sandbox Management
   - [x] [P0]RunPodSandbox #455 
   - [x] [P0]StopPodSandbox #561 
   - [x] [P0]RemovePodSandbox #561 
   - [x] [P0] ListPodSandbox #551 
   - [x] [P0] PodSandboxStatus #573 

- [x] Container Management
   - [x] [P0] CreateContainer #487 
   - [x] [P0] StartContainer #487 
   - [x] [P0] StopContainer #527 
   - [x] [P0] RemoveContainer #527 
   - [x] [P0] ListContainers #527 
   - [x] [P0] ContainerStatus #573 

- [x] Image Management
   - [x] [P0] ListImages #477 
   - [x] [P0] PullImages #477 
   - [x] [P0] RemoveImages #496 
   - [x] [P0] ImageStatus #496 
   - [x] [P2] ImageFsInfo

### Others
- [x] Network
   - [x] [P1] Integrate CNI plugin #580 

- [x] Streaming
   - [x] [P0] ExecSync #629 
   - [x] [P2] Exec #797 
   - [x] [P2] Attach #797 
   - [x] [P2] PortForward #816

- [x] [P2] Logging #928 

cc @skoo87 @allencloud

### PR DESCRIPTION
**Pull request by YaoZengzeng**
**1.Describe what this PR did**

This PR implement the skeleton of cri manager which includes:

* framework of cri manager

* grpc server

* the definition of methods which we need to handle cri requests

**2.Does this pull request fix one issue?** 

fix #420 

**3.Describe how to verify it**

Maybe the [cri-tools](https://github.com/kubernetes-incubator/cri-tools) is what you need.

We may use the following method to verify cri manager is really work:
```
[root@pouch-test monster]# crictl info
DEBU[0000] StatusRequest: &StatusRequest{Verbose:true,} 
DEBU[0000] StatusResponse: nil                          
FATA[0000] getting status of runtime failed: rpc error: code = Unknown desc = Status Not Implemented Yet  
```

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -25,6 +26,7 @@ type Daemon struct {
 	systemMgr      mgr.SystemMgr
 	imageMgr       mgr.ImageMgr
 	volumeMgr      mgr.VolumeMgr
+	criManager     *cri.CRIManager
```
Comment (allencloud) [inline daemon/daemon.go]: To keep consistent, criMgr and CRIMgr?

File: daemon/daemon.go [hunk 2]
```diff
@@ -102,7 +110,28 @@ func (d *Daemon) Run() error {
 		VolumeMgr:    volumeMgr,
 	}
 
-	return d.server.Start()
+	httpServerCloseCh := make(chan struct{})
+	go func() {
+		if err := d.server.Start(); err != nil {
+			logrus.Errorf("Failed to start http server: %v", err)
```
Comment (allencloud) [inline daemon/daemon.go]: please lowercase the beginning letter in log if not necessary.

File: daemon/daemon.go [hunk 3]
```diff
@@ -102,7 +110,28 @@ func (d *Daemon) Run() error {
...
+		}
+		close(httpServerCloseCh)
+	}()
+
+	grpcServerCloseCh := make(chan struct{})
+	go func() {
+		if err := d.criManager.Serve(); err != nil {
+			logrus.Errorf("Failed to start grpc server: %v", err)
```
Comment (allencloud) [inline daemon/daemon.go]: The same as above.

File: daemon/daemon.go (hunk 1 above)
Comment (YaoZengzeng) [reply daemon/daemon.go]: Actually, in other cases, the type of *Manager is struct and *Mgr is interface.

From this perspective, "CRIManager" is more consistent 😄 , but "CRIManager" seems can't pass go lint which said it is too stutter.

I'll change it to criMgr.

**Comment by allencloud** [comment]
Please rebase to the latest master, since PR https://github.com/alibaba/pouch/pull/436 which solved flaky test has been merged. Thanks.

**Comment by allencloud** [comment]
Please rebase to the latest master, since PR https://github.com/alibaba/pouch/pull/436 which solved flaky test has been merged. Thanks.

**Comment by allencloud** [comment]
At the very beginning, I think we need to consider how to provide this kind of ability to our users. 
How to give an overall picture of CRI implementation, how to construct our API docs, how to provide installation guide, how to guarantee the quality of this part and so on. Maybe a plan on all these aspects is necessary. @YaoZengzeng @zzchun @zeppp @ZouRui89

**Comment by YaoZengzeng** [comment]
@allencloud I totally agree with you.

Maybe we should use this comment to supplement issue #420 which will guide our work . 😄

File: cri/cri.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	// TODO: Prepare streaming server.
+
+	runtime.RegisterRuntimeServiceServer(s.server, s.criMgr)
+	runtime.RegisterImageServiceServer(s.server, s.criMgr)
+
+	return s, nil
+}
+
+// Serve starts grpc server.
+func (s *Service) Serve() error {
```
Comment (skoo87) [inline cri/cri.go]: How to listen on TCP ????

File: daemon/config/config.go [hunk 5]
```diff
@@ -17,6 +17,9 @@ type Config struct {
 	// Server listening address.
 	Listen []string
 
+	// ListenCri is the listening address which serves CRI.
+	ListenCri string
```
Comment (skoo87) [inline daemon/config/config.go]: ListenCri  -> ListenCRI

File: cri/cri.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+package cri
+
+import (
+	"net"
+	"os"
+	"syscall"
+
+	"google.golang.org/grpc"
```
Comment (skoo87) [inline cri/cri.go]: put "google.golang.org/grpc" and "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime" together.

**Comment by YaoZengzeng** [comment]
@skoo87 updated!

File: daemon/mgr/cri.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
+package mgr
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
```
Comment (skoo87) [inline daemon/mgr/cri.go]: use "context" in standard library to replace "golang.org/x/net/context".

**Comment by skoo87** [comment]
LGTM

**Comment by allencloud** [comment]
LGTM, thanks for your work.

**Pull request merged**